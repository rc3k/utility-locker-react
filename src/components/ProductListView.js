import React from 'react';
import { withRouter, Link } from 'react-router-dom';
import { connect } from 'react-redux';
import * as actions from '../actions/collection';

export class ProductListView extends React.Component {
  componentDidMount() {
    const { loadCollection } = this.props;
    loadCollection('products');
  }

  render() {
    const { items, meta } = this.props;

    if (meta.status === 'REQUESTED') {
      return (
        <div className="loader">Loading...</div>
      )
    }

    return (
      <div className="collection">
        <h1>{meta.collectionName}</h1>
        <table>
          <thead>
            <tr>
              <th>Name</th>
            </tr>
          </thead>
          <tbody>
            {items.map((item) => (
              <tr key={item.id}>
                <td>
                  <Link to={`/${item.id}`}>{item.name}</Link>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

export const ProductListViewContainer = withRouter(
  connect((state, ownProps) => ({
    items: state.products.items,
    meta: state.products.meta,
  }), actions)(ProductListView),
);
