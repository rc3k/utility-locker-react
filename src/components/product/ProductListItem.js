import React from 'react';

export default ({ item }) => (
  <div className="product-list-item list-group-item flex-column align-items-start mb-2">
    <div className="d-flex w-100 justify-content-between">
      <h5 className="mb-1">{item.name}</h5>
      <div>
        Â£
        {item.price.value.toFixed(2)}
      </div>
    </div>
    <p className="mb-1">{item.description}</p>
    <small>
      <ul>
        <li>
          <span className="font-weight-bold">ID: </span>
          {item.id}
        </li>
        <li>
          <span className="font-weight-bold">Type: </span>
          {item.type_name}
        </li>
        <li>
          <span className="font-weight-bold">Department: </span>
          {item.department}
        </li>
        <li>
          <span className="font-weight-bold">Weight: </span>
          {item.weight}
          g
        </li>
      </ul>
    </small>
  </div>
);
