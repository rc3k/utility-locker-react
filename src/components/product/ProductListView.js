import React from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { products } from '../../slices/products';
import { loadProducts } from '../../slices/thunks';
import ProductListItem from "./ProductListItem";
import Preloader from "../Preloader";
import Header from "../Header";

export class ProductListView extends React.Component {
  componentDidMount() {
    const { loadProducts } = this.props;
    loadProducts({ type: '' });
  }

  render() {
    const { products, loadProducts, history } = this.props;
    const { items, loading, params } = products;

    if (loading) {
      return (
        <Preloader />
      )
    }

    return (
      <div className="list-view">
        <Header
          heading="Products"
          setSort={(column, direction) => loadProducts({ column, direction })}
          sortParams={params}
          groupByField=""
          setGroupBy={groupBy => history.push(`/${groupBy}`)}
        />
        <div className="list-group mt-2">
          {items.allIds.map(itemId => <ProductListItem key={itemId} item={items.byId[itemId]} />)}
        </div>
      </div>
    );
  }
}

export const ProductListViewContainer = withRouter(
  connect((state, ownProps) => ({
    products: state.products
  }), { ...products.actions, ...{ loadProducts } })(ProductListView),
);
