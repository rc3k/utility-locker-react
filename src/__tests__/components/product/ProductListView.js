import React from 'react';
import renderer from 'react-test-renderer';

import { BrowserRouter as Router } from 'react-router-dom';
import { ProductListView } from '../../../components/product/ProductListView';

it('renders correctly when data is loading', () => {
  const listView = renderer.create(
    <Router>
      <ProductListView
        products={{loading: true}}
        loadProducts={() => {}}
      />
    </Router>,
  ).toJSON();
  expect(listView).toMatchSnapshot();
});

it('renders correctly when items have loaded', () => {

  const listView = renderer.create(
    <Router>
      <ProductListView
        products={
          {
            loading: false,
            items: {
              byId: {
                '123-X': {
                  id: '123-X',
                  name: 'Item one',
                  description: 'Item description',
                  weight: 123,
                  price: {
                    value: 156.98
                  },
                  type_name: 'Book',
                },
                '124-Y': {
                  id: '123-Y',
                  name: 'Item two',
                  description: 'Item description 2',
                  weight: 987,
                  price: {
                    value: 1.23
                  },
                  type_name: 'Electrical',
                },
              },
              allIds: ['123-X', '124-Y',],
            },
            params: {
              column: 'name',
              direction: 'asc'
            },
          }
        }
        loadProducts={() => {}}
      />
    </Router>,
  ).toJSON();
  expect(listView).toMatchSnapshot();
});
