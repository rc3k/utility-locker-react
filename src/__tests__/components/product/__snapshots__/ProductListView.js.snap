// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when data is loading 1`] = `
<div
  className="loader"
>
  Loading...
</div>
`;

exports[`renders correctly when items have loaded 1`] = `
<div
  className="list-view"
>
  <div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <h1>
        Products
      </h1>
    </div>
    <div
      className="col-md-6"
    >
      <div
        className="d-flex w-100 p-2"
      >
        <label
          className="pt-2 pr-1 text-right"
          htmlFor="sort-by-select"
        >
          Sort:
        </label>
        <select
          className="form-control"
          id="sort-by-select"
          onChange={[Function]}
          value="name_asc"
        >
          <option
            value="name_asc"
          >
            Default (name)
          </option>
          <option
            value="price_desc"
          >
            Highest price
          </option>
          <option
            value="price_asc"
          >
            Lowest price
          </option>
        </select>
      </div>
      <div
        className="d-flex w-100 form-group p-2"
      >
        <label
          className="pt-2 pr-1 text-right"
          htmlFor="group-by-select"
        >
          Group by:
        </label>
        <select
          className="form-control"
          id="group-by-select"
          onChange={[Function]}
          value=""
        >
          <option
            value=""
          >
            None
          </option>
          <option
            value="type"
          >
            Type
          </option>
        </select>
      </div>
    </div>
  </div>
  <div
    className="list-group mt-2"
  >
    <div
      className="product-list-item list-group-item flex-column align-items-start mb-2"
    >
      <div
        className="d-flex w-100 justify-content-between"
      >
        <h5
          className="mb-1"
        >
          Item one
        </h5>
        <div>
          £
          156.98
        </div>
      </div>
      <p
        className="mb-1"
      >
        Item description
      </p>
      <small>
        <ul>
          <li>
            <span
              className="font-weight-bold"
            >
              ID: 
            </span>
            123-X
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Type: 
            </span>
            Book
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Department: 
            </span>
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Weight: 
            </span>
            123
            g
          </li>
        </ul>
      </small>
    </div>
    <div
      className="product-list-item list-group-item flex-column align-items-start mb-2"
    >
      <div
        className="d-flex w-100 justify-content-between"
      >
        <h5
          className="mb-1"
        >
          Item two
        </h5>
        <div>
          £
          1.23
        </div>
      </div>
      <p
        className="mb-1"
      >
        Item description 2
      </p>
      <small>
        <ul>
          <li>
            <span
              className="font-weight-bold"
            >
              ID: 
            </span>
            123-Y
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Type: 
            </span>
            Electrical
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Department: 
            </span>
          </li>
          <li>
            <span
              className="font-weight-bold"
            >
              Weight: 
            </span>
            987
            g
          </li>
        </ul>
      </small>
    </div>
  </div>
</div>
`;
